Partially Fixes issue #6903 ( Prepare>Data Reshape>Random Subset )

EstherNjeriLiberatta opened the pull request:
Fixes #6903 In this PR I have made the requested changes as suggested by @lilyclements .

@rdstern @lilyclements Kindly review.
@africanmathsinitiative/developers This is ready for review.

lloyddewit commented:
@Ivanluv Please could you review? thanks

lloyddewit commented on instat/dlgRandomSubsets.vb:
```diff
@@ -18,7 +18,7 @@ Imports instat.Translations
     Public bFirstLoad As Boolean = True
     Private bReset As Boolean = True
-    Private clsDataFrame, clsSetSeed, clsSample, clsReplicate As New RFunction
+    Private clsDataFrameFunction, clsSampleNFunctionDataframeFunction, clsSetSeed, clsSampleFunction, clsRerunFunction As New RFunction
```
Thanks for naming these `...Function`, this helps readability. 
Should we also rename `clsSetSeed` to `clsSetSeedFunction` for consistency?

lloyddewit reviewed (changes requested):
@EstherNjeriLiberatta Looks good, just one small comment

EstherNjeriLiberatta commented:
@lloyddewit I have made the requested changes. Thanks

lloyddewit commented:
@rdstern please could you test? thanks

rdstern commented:
![image](https://user-images.githubusercontent.com/11226469/150380119-3d58c537-5373-4f0d-ad8f-72891636ab02.png)

a) Please check the code.  You can see I don't get 10 columns here, just 2.  Ask @HawardKetoyoMsatsi or others if you need help.
b) I used the checkbox to set the seed.  When I return to the dialogue, it is unchecked.  It should remember it was checked before.  The other controls do all seem to remember.

rdstern reviewed (changes requested):
Comments given for changes needed

EstherNjeriLiberatta commented:
@rdstern I have fixed the changes you requested. Kindly review.

@africanmathsinitiative/developers This is ready for review.

rdstern reviewed (dismissed):
@EstherNjeriLiberatta great.  This now seems to work fine.

shadrackkibet commented on instat/dlgRandomSubsets.vb:
```diff
@@ -130,7 +142,7 @@ Public Class dlgRandomSubsets
 
     Private Sub ucrBase_BeforeClickOk(sender As Object, e As EventArgs) Handles ucrBase.BeforeClickOk
         If ucrChkSetSeed.Checked Then
-            frmMain.clsRLink.RunScript(clsSetSeed.ToScript(), strComment:="dlgRandomSubset: Setting the seed for random number generator")
+            frmMain.clsRLink.RunScript(clsSetSeedFunction.ToScript(), strComment:="dlgRandomSubset: Setting the seed for random number generator")
         End If
     End Sub
```
This should change. We should run this when set seed checkbox state changes i.e ....controlValue changed event. Then we should do this as before codes. So something like this.
```
    Private Sub ucrChkSetSeed_ControlValueChanged(ucrChangedControl As ucrCore) Handles ucrChkSetSeed.ControlValueChanged
        If ucrChkSetSeed.Checked Then
            ucrBase.clsRsyntax.AddToBeforeCodes()
        Else
            ucrBase.clsRsyntax.RemoveFromBeforeCodes()
        End If
    End Sub
```

Remember to clear codes in setdefault.

shadrackkibet commented on instat/dlgRandomSubsets.vb:
```diff
@@ -70,46 +71,57 @@ Public Class dlgRandomSubsets
-        clsReplicate.AddParameter("n", 1)
-        clsReplicate.AddParameter("expr", clsRFunctionParameter:=clsSample)
+        'sample_n dataframe func setting
+        clsSampleNFunctionDataframeFunction.SetRCommand("data.frame")
```
Remove this function because we can force `get_columns_from_data()` to extract a single column as a dataframe by doing
`ucrReceiverSubsets.bForceAsDataFrame = True` on initialise.

shadrackkibet reviewed (changes requested):
@EstherNjeriLiberatta well done. This is good progress.

I have reviewed your PR by adding in comments and suggestions where possible. I hope these comments are useful to improve the code. Let me know if something isn't clear enough. I am happy to write more here to explain if you feel something isn't clear or even follow up elsewhere.

EstherNjeriLiberatta commented:
@africanmathsinitiative/developers This is ready for review.

shadrackkibet commented on instat/dlgRandomSubsets.vb:
```diff
@@ -184,9 +177,15 @@ Public Class dlgRandomSubsets
+
+    Private Sub ucrBase_BeforeClickOk(sender As Object, e As EventArgs) Handles ucrBase.BeforeClickOk
+        If ucrChkSetSeed.Checked Then
+            frmMain.clsRLink.RunScript(clsSetSeedFunction.ToScript(), strComment:="dlgRandomSubset: Setting the seed for random number generator")
```
Do you still need this? I see it already in line 180.

EstherNjeriLiberatta commented:
@shadrackkibet Thank you for the corrections. 
@africanmathsinitiative/developers This is ready for review.

rdstern reviewed (dismissed):
Still works ok for me

shadrackkibet commented on instat/dlgRandomSubsets.vb:
```diff
@@ -70,46 +72,54 @@ Public Class dlgRandomSubsets
+        'rerun func setting
+        clsRerunFunction.SetPackageName("purrr")
+        clsRerunFunction.SetRCommand("rerun")
+        clsRerunFunction.AddParameter(".n", 1)
+        clsRerunFunction.AddParameter("expr", clsRFunctionParameter:=clsSampleFunction)
```
With parameter names, the position doesn't matter. For consistency, always add parameter positions. Check other places as well.

EstherNjeriLiberatta replied:
Alright, thanks.

shadrackkibet commented on instat/dlgRandomSubsets.vb:
```diff
@@ -169,4 +174,12 @@ Public Class dlgRandomSubsets
     Private Sub CoreControls_ControlContentsChanged(ucrChangedControl As ucrCore) Handles ucrNudNumberOfColumns.ControlContentsChanged, ucrNudSampleSize.ControlContentsChanged, ucrNudSetSeed.ControlContentsChanged, ucrNewDataFrame.ControlContentsChanged, ucrChkSetSeed.ControlContentsChanged, ucrReceiverSubsets.ControlContentsChanged
         TestOKEnabled()
     End Sub
```
I like to make this the last `sub`

EstherNjeriLiberatta replied:
Okay

shadrackkibet commented on instat/dlgRandomSubsets.vb:
```diff
@@ -37,7 +37,8 @@ Public Class dlgRandomSubsets
 
         'ucrReceiver
-        ucrReceiverSubsets.SetParameter(New RParameter("x", 2))
+        ucrReceiverSubsets.bForceAsDataFrame = True
```
Since you now have this. Should just remove `clsSampleNFunctionDataframeFunction`

EstherNjeriLiberatta replied:
I need that function to convert the variable in the sample_n function into a data frame that is;
`sample_n (data.frame(<variable>), size=...)`. It didn't work when I removed it.

I also decided to keep `bIncludeArgumentName:= False` in line 40 as it is because I did not want `x`  to appear within the function e.g. `sample_n (data.frame(x=field), size=...)`. I would like the R code to look simple. Thanks.

shadrackkibet replied:
You mention that " It didn't work when I removed it". What exactly do you mean? Did it give an error? If yes which error message? Did you get an exception etc... 

Just saying that things didn't work doesn't help me to help you.

EstherNjeriLiberatta replied:
Okay, sorry. I guess I can show you what the R code looks like. I didn't know how best to explain it.
The function is on the second highlight. Thanks for helping.

![image](https://user-images.githubusercontent.com/74548350/151523187-a53e04f7-f8dd-4d2d-84e6-3c5514d24ade.png)

shadrackkibet replied:
Does this work for you?
```

# Code generated by the dialog, Random Subsets

field <- data_book$get_columns_from_data(data_name="survey", col_names="field", force_as_data_frame=TRUE)
survey_random <- data.frame(X=purrr::rerun(.n=1, expr=dplyr::sample_n(tbl=field, size=36)))
data_book$import_data(data_tables=list(survey_random=survey_random))

rm(list=c("survey_random", "field"))
```

EstherNjeriLiberatta replied:
No, the field variable has to be converted into a data frame in the function `sample_n()`

EstherNjeriLiberatta replied:
It should be `sample_n(tbl=data.frame(<variable>),...`

shadrackkibet replied:
Isn't it already a data frame when you do this? `field <- data_book$get_columns_from_data(data_name="survey", col_names="field", force_as_data_frame=TRUE)`

EstherNjeriLiberatta replied:
I get this error when I try that. 
```
field <- data_book$get_columns_from_data(data_name="survey", col_names="field", col_names="field", force_as_data_frame=TRUE)
survey_random2 <-data.frame(X=purrr::rerun(.n=4, expr=dplyr::sample_n(tbl=field, size=36)))
data_book$import_data(data_tables=list(survey_random2=survey_random2))
```

![image](https://user-images.githubusercontent.com/74548350/151542251-4d1b2e6c-d8e6-4ccb-aad7-769a75540d88.png)

shadrackkibet replied:
That is because you have `col_names` twice in this line; 

`data_book$get_columns_from_data(data_name="survey", col_names="field", col_names="field", force_as_data_frame=TRUE)`

EstherNjeriLiberatta replied:
May I contact you on skype, if you don't mind

shadrackkibet replied:
Sure

shadrackkibet replied:
Resolved over skype.

EstherNjeriLiberatta commented:
I made corrections to the code:
i. Removed `clsSampleNDataFrameFunction` as it was repeating what `ucrReceiverSubsets.bForceAsDataFrame = True` was doing i.e. converting the varibles into a data frame.
ii. Added parameter positions to all parameters for consistency.

@shadrackkibet Thank you.
@africanmathsinitiative/developers This is ready for review.

shadrackkibet commented:
@rdstern could you test and re-approve? In theory, there shouldn't be any functionality change since your last approval.

@EstherNjeriLiberatta just to let you know that I made a small edit in the `clsRSyntax.vb` file.

![image](https://user-images.githubusercontent.com/26301973/151768983-8a7833c3-6d03-46bd-b7b9-7eb56b59f1db.png)

rdstern reviewed (approved):
Well done.  Seems fine to me
